### PR TITLE
Meterpreter python extension

### PR DESCRIPTION
**Pull request by OJ**
This PR is the Metasploit-side of the payloads PR located here https://github.com/rapid7/metasploit-payloads/pull/37

This is where I will document and discuss all the detail of the implementation, so that it can all be done in one place.
## Overview

The goal of this extension is to provide the means for people to easily script and run functionality on the target machine without touching the file system. This extension provides an in-memory implementation of Python with a bunch of base libraries that will be needed for most cases. This includes sys, os, threads, cTypes, cPickle, etc. The cTypes is particularly interesting to us because it means we can do more Windowsy stuff on the client.

The ultimate goal is to also be able to work on controller Meterpreter functionality itself, and being able to call certain functions within it (thing of something like `meterpreter.getsystem()` for example). This functionality isn't ready to be PR'd yet, but it's coming soon.
## Current features
- Windows only I'm afraid
- Currently runs on both x86 and x64
- Runs in memory
- Is a full RDI DLL built from the Python source
- Should build without a problem with the same toolset that we use to build all the other extensions, however it's not _clean_ because building Python from source doesn't work without warnings anyway
- Modules can be imported, file by file or by folder using `python_import`
- Arbitrary python code can be run in the context of the session, line by line if required, by using `python_execute`
- The calls to the Python interpreter are persistent, so if a module is imported it stays there
- If the user wishes to reset the state of the interpreter they can use the `python_reset` command
## Implementation detail

This is a little bit hairy.

The DLL is an RDI binary, much the same as the rest of them, with Python's source built directly. The python interpreter is not pre-built and stored as a resource like some other implementations. Most of the built-in stuff that comes in .pyd files is actually built into this extension, and this means we don't need to have other libs dynamically loaded. It does also mean that the binary is a bit bigger. It was a tradeoff I was happy to make. Happy to discuss this further.

The built-in libraries that are written in python needed to be shipped with it. To support this, and to support reading modules from memory instead of from disk, a custom loader exists in the code that resolves modules from memory. This module, along with all the system libs, are compiled into pyc files, compressed, and baked into the extension as a resource. This resource is called `python_core.cz`, and is built into the library at compile time.

Thanks to the nature of RDI binaries not registering themselves as a module with the host process, using the `*Resource` Win32 functions was painful because of the lack of a module handle. However, there's a common trick used to get the module handle based on memory addresses, and so that was used instead. With that, I was able to load resources in memory, wire them into the interpreter, and have the python code executed so that the modules are present. This obviously needs to be reviewed and tested, and critiqued!

PYC files compiled on linux don't appear to work on Windows. Hence, they are pre-built and included in `python_core.cz`. Happy to discuss this further.

So the process goes:
- User types `use python`.
- `ext_server_python.<arch>.dll` is pushed up to the session just like all the other binaries.
- The extension initialisation function gets called.
- The code initialises the python run time.
- Lots of internal functions are wired up.
- The `python_core.cz` file is located, and referenced in memory.
- The content of the compressed resource is passed to python so that it can be loaded/accessed.
- Python code is run to wire up the appropriate functions so that modules can be loaded from memory.
- The compressed libraries are loaded and ready for importing.

From there we're ready to go.

There is a helper script for creating `python_core.cz`, and at this point in time it has to be run on a Windows machine.
## Things left to do prior to this PR landing
- [x] Iron out any creases in managing of Python objects in C land (not sure if I have correctly covered all cases here).
- [x] Make sure I include the licence for Python in the appropriate way (ping @todb-r7 for help here please).
## Things left to do after this PR is landed
- Finalise the approach for the Meterpreter bindings. This will still take a bit of work, as some underlying "stuff" will need to change, but it shouldn't be _too_ drastic.
- Implement Meterpreter bindings for all the functions that are available. This is going to take time to expose the lot, unless someone can come up with something crafty to get around it (not sure this is possible though).
- Wire in the existing libre SSL library so that it can be reached from within Python. This is going to take a little bit of work, but I have a path forward for this, and this will mean that we don't need to push up another copy of SSL with the extension. Currently, no SSL support is baked in.
- Take the zlib bit out of the python lib, and wire in the existing zlib stuff that comes with metsrv. This will have to take the same approach as the SSL stuff, but it was needed quickly to make all the things work properly and hence this functionality is duplicated a bit as a result.
- Spend some time reducing the size. There are probably lots of things that can be removed, but I'm not sure what they are at this point.
## Testing

I've run this on Windows XP x86, Windows 7 x64, Windows 8.1 x64, Windows 2012 x64 and Windows 10 x64. It all "seemed to work ok" :) worksonmymachines.gif
## Sample output

Shown from Windows 7 x64:

```
meterpreter > sysinfo
Computer        : WIN-S45GUQ5KGVK
OS              : Windows 7 (Build 7601, Service Pack 1).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/win64
meterpreter > use python
Loading extension python...success.
meterpreter > help

... snip ...

Python Commands
===============

    Command         Description
    -------         -----------
    python_execute  Execute a python command string
    python_import   Import/run a python file or module
    python_reset    Resets/restarts the Python interpreter
```

Sample direct run of a bit of python code:

```
meterpreter > python_execute "print 'hi from python!'"
[+] Content written to stdout:
hi from python!
```

Another example, importing a module and retrieving the content of a variable as a string:

```
meterpreter > python_execute "import os; cd = os.getcwd()" -r cd
[+] cd = C:\Users\OJ\Desktop
```

Importing and running a local module

```
meterpreter > python_import -f /tmp/enumwindows.py
[*] Importing /tmp/enumwindows.py ...
[+] Content written to stdout:
[u'Start', u'', u'DebugView on \\\\WIN-S45GUQ5KGVK (local)', u'Developer Command Prompt for VS2013', u'Lib', u'meterpreter - Microsoft Visual Studio', u'Program Manager']
```

A folder can be used as a module path, but there are a few things around this that need work. I'd like to call on the likes of @zeroSteiner to help me here.

This shows that reset works by clearing our imported modules

```
meterpreter > python_execute "import os,ctypes,cPickle"
[+] Command executed without returning a result
meterpreter > python_execute "print dir()"
[+] Content written to stdout:
['MetFinder', 'MetLoader', '__builtins__', '__doc__', '__name__', '__package__', 'cPickle', 'ctypes', 'imp', 'marshal', 'met_dbg', 'met_dbg_trace', 'met_finder', 'met_import_code', 'met_init', 'met_mod_body', 'met_mod_name', 'os', 'sys']

meterpreter > python_reset
[+] Python interpreter successfully reset
meterpreter > python_execute "print dir()"
[+] Content written to stdout:
['MetFinder', 'MetLoader', '__builtins__', '__doc__', '__name__', '__package__', 'imp', 'marshal', 'met_dbg', 'met_dbg_trace', 'met_finder', 'met_import_code', 'met_init', 'met_mod_body', 'met_mod_name', 'sys']
```

The rest I would like to leave up to your imagination. Much more to come here though! Thanks for the eyes, support etc. I look forward to getting this in and refining it.
## Verification
- [ ] Get x86 and x64 sessions
- [ ] make sure `use python` works in both
- [ ] make sure `python_execute` works in both
- [ ] make sure `python_import` works (well enough) in both
- [ ] make sure `python_reset` works in both

### CODE REVIEW

**Comment by todb-r7** [comment]
With regard to the license material for Python, it looks like it's pretty straightforward to include the PSF license in the regular LICENSE file here in metasploit-framework. It doesn't look like the metasploit-payloads repo has any license info at all, but that's pretty easily remedied by dropping in an identical COPYING file there and a similarly machine-readable LICENSE file.

I can sort that on Monday, @OJ, in the form of a pull request to rapid7/metasploit-payloads#37

**Comment by todb-r7** [comment]
In any event, the PSF license looks plenty permissive and compatible with the normal licensing of Metasploit.

**Comment by OJ** [comment]
Thanks @todb-r7 ! Appreciate the help mate.

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/python.rb [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+  @@python_import_opts = Rex::Parser::Arguments.new(
+    '-h' => [false, 'Help banner'],
+    '-f' => [true,  'Path to the file (.py, .pyc), or module directory to import'],
+    '-n' => [true,  'Name of the module (optional, for single files only)'],
+    '-r' => [true,  'Name of the variable containing the result (optional, single files only)']
+  )
+
+  def python_import_usage
+    print_line('Usage: python_imoprt <-f file path> [-n mod name] [-r result var name]')
```
Comment (wvu-r7) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/python.rb]: ```python_imoprt``` should be ```python_import```.

**Comment by OJ** [comment]
Thanks @wvu-r7!

**Comment by all3g** [comment]
Excellent job. Thanks so much.

**Comment by hmoore-r7** [comment]
Looks great! One small missing feature would be the ability to execute a specific blob of python code during initialization (in a stageless meterpreter context).

**Comment by todb-r7** [comment]
@OJ, see https://github.com/OJ/metasploit-payloads/pull/1 to pick up license details for Python.

**Comment by byt3bl33d3r** [comment]
@OJ this is awesome! Can't wait to play with this :D

**Comment by zeroSteiner** [comment]
I was able to build and run this without any issues. It doesn't look like if a folder is used as a module that the files in it can import each other (maybe I'm not doing it right). Is the intent to be able to run a largish package with a \_\_main\_\_.py script, to make the module available to subsequent calls to `python_execute` or maybe something else?

On some of the special python variables:
`__file__` seems to be the module name plus .py instead of the name of a file. According to the [docs](https://docs.python.org/2/reference/datamodel.html) I think it'd be safe to just remove `__file__` altogether since it can't point to a legit path on disk. `__name__` is correctly set to `__main__` when using `python_execute` but **not** when using `python_import -f`, which is returning `hello_world.__main__` in the case of a file `hello_world/__main__.py`.

Also props for preventing `sys.exit` from killing the session, could the same technique also be applied to `os._exit` also to prevent accidental session deaths? This is awesome, great job @OJ!

**Comment by OJ** [comment]
Thanks for the great feedback folks.

@hmoore-r7: the stageless thing will come. I had that as a lower priority than the rest at this point.

@zeroSteiner: Great feedback, thanks! I'll chat to you on IRC about the special vars, as I knew I'd have them wrong along the way. Also, will talk more about the resolver reaching back to MSF and getting things on the fly like you suggested.

Latest news: https://gist.github.com/OJ/9964bf9d7724869256f5 Bindings are coming together.

**Comment by OJ** [comment]
Thanks Brent!

**Pull request merged**